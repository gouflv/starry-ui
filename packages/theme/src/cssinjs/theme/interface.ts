// cssinjs is reference to @ant-design/cssinjs that use isolated types

export type TokenType = object
export type DerivativeFunc<
  DesignToken extends TokenType,
  DerivativeToken extends TokenType
> = (
  designToken: DesignToken,
  derivativeToken?: DerivativeToken
) => DerivativeToken
