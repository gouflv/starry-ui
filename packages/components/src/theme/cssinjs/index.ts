export { default as DesignTokenProvider } from './DesignTokenProvider'
export { default as useToken } from './useToken'
