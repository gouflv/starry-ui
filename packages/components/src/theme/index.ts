export { css, cx } from '@emotion/css'
export * from './style'
