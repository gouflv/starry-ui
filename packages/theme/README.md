# Starry UI Theme

## Overview

Flow of token generation:

```
DerivativeFunc --> Theme

SeedToken --> Theme --> MapToken -> AliasFormat -> AliasToken
```
