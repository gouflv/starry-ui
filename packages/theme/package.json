{
  "name": "@starry-ui/theme",
  "description": "Starry UI theme package",
  "version": "1.0.5",
  "license": "MIT",
  "homepage": "https://starry-ui.netlify.app",
  "main": "dist/index.js",
  "module": "dist/index.mjs",
  "types": "dist/index.d.ts",
  "repository": {
    "type": "git",
    "url": "https://github.com/gouflv/starry-ui"
  },
  "keywords": [
    "css-in-js",
    "starry-ui",
    "theme",
    "ui",
    "vue"
  ],
  "files": [
    "dist"
  ],
  "scripts": {
    "build": "tsup",
    "token-meta": "node scripts/generate-token-meta.js",
    "prepublishOnly": "pnpm build",
    "pub": "pnpm publish --access=public"
  },
  "devDependencies": {
    "@tsconfig/node20": "^20.1.4",
    "@types/node": "^22.7.7",
    "@vue/test-utils": "^2.4.6",
    "fs-extra": "^11.2.0",
    "tsup": "^8.3.0",
    "typedoc": "^0.23.28",
    "typescript": "^5.6.3",
    "vue": "^3.5.12"
  },
  "dependencies": {
    "@ant-design/colors": "^7.1.0",
    "@ctrl/tinycolor": "^4.1.0"
  },
  "peerDependencies": {
    "vue": ">=3.2"
  }
}
