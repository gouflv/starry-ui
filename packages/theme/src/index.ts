import { TinyColor } from '@ctrl/tinycolor'
import compactAlgoliaTheme from './themes/compact'
import defaultAlgoliaTheme from './themes/default'

export * from './cssinjs'
export * from './interface'
export * from './internal'
export { TinyColor }

export const themes = {
  default: defaultAlgoliaTheme,
  compact: compactAlgoliaTheme,
}
