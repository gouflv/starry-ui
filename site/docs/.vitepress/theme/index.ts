import 'ant-design-vue/dist/reset.css'
import { createPinia } from 'pinia'
import 'virtual:uno.css'
import DefaultTheme from 'vitepress/theme'
import Layout from './Layout.vue'
import './main.css'

export default {
  extends: DefaultTheme,
  Layout,
  async enhanceApp({
    //@ts-ignore
    app
  }) {
    const pinia = createPinia()
    app.use(pinia)

    if (!import.meta.env.SSR) {
      const AntConfig = await import('@/components/common/AntConfig.vue')
      app.component('AntConfig', AntConfig.default)

      const ThemeEditor = await import('@/components/theme-editor/index.vue')
      app.component('ThemeEditor', ThemeEditor.default)
    }
  }
}
