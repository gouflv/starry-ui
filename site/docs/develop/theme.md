# 主题

**主题** 是用于统一界面风格和视觉效果的配置集合。 其中 Design Token 是设计系统中用来存储视觉样式的最小单位，包括颜色、字体大小、边距、动画等。

在 Starry UI 中，我们将主题抽象到一个独立的主题库 `@starry/theme` 中，以便于在不同的平台和框架中使用。 主题库提供了一套完整的主题方案，包括 Token、梯度算法、模式算法和 CSS-in-JS 运行时方案等，使得开发人员可以快速使用和定制主题。

## Design Token

我们将 Design Token 抽象成三层结构，分别是 Seed Token、Map Token 和 Alias Token。

::: tip
在大部分情况下，使用 Seed Token 就可以满足定制主题的需要。
:::

### 基础变量（Seed Token）

| Token 名称      | 描述                                           | 类型    | 默认值                                                                                                                                                                                |
| --------------- | ---------------------------------------------- | ------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| borderRadius    | 基础组件的圆角大小，如按钮、输入框等           | number  | 6                                                                                                                                                                                     |
| colorBgBase     | 用于背景色梯度的基础变量，生成背景色梯度变量   | string  | #fff                                                                                                                                                                                  |
| colorError      | 表示操作失败的颜色，如错误状态提示             | string  | #ff4d4f                                                                                                                                                                               |
| colorInfo       | 表示操作信息的颜色，如 Alert、Tag 等           | string  | #1677ff                                                                                                                                                                               |
| colorLink       | 控制超链接的颜色                               | string  | #1677ff                                                                                                                                                                               |
| colorPrimary    | 品牌色                                         | string  | #1677ff                                                                                                                                                                               |
| colorSuccess    | 表示操作成功的颜色，如 Result、Progress 等组件 | string  | #52c41a                                                                                                                                                                               |
| colorTextBase   | 用于文本色梯度的基础变量，生成文本色梯度变量   | string  | #000                                                                                                                                                                                  |
| colorWarning    | 表示操作警告的颜色，如 Notification 等         | string  | #faad14                                                                                                                                                                               |
| controlHeight   | 基础控件的高度，如按钮和输入框                 | number  | 32                                                                                                                                                                                    |
| fontFamily      | 使用系统默认界面字体，提供屏显友好的备用字体   | string  | -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, 'Helvetica Neue', Arial, 'Noto Sans', sans-serif, 'Apple Color Emoji', 'Segoe UI Emoji', 'Segoe UI Symbol', 'Noto Color Emoji' |
| fontFamilyCode  | 代码字体，用于 code、pre、kbd 元素             | string  | 'SFMono-Regular', Consolas, 'Liberation Mono', Menlo, Courier, monospace                                                                                                              |
| fontSize        | 常用的字体大小                                 | number  | 14                                                                                                                                                                                    |
| lineType        | 控制边框和分割线样式                           | string  | solid                                                                                                                                                                                 |
| lineWidth       | 控制边框和分割线宽度                           | number  | 1                                                                                                                                                                                     |
| motion          | 配置动画效果，`false` 时关闭动画               | boolean | true                                                                                                                                                                                  |
| sizePopupArrow  | 组件箭头的尺寸                                 | number  | 16                                                                                                                                                                                    |
| sizeStep        | 控制组件尺寸的基础步长                         | number  | 4                                                                                                                                                                                     |
| sizeUnit        | 控制尺寸变化单位                               | number  | 4                                                                                                                                                                                     |
| zIndexBase      | 组件基础 Z 轴值                                | number  | 0                                                                                                                                                                                     |
| zIndexPopupBase | 浮层类组件的基础 Z 轴值                        | number  | 1000                                                                                                                                                                                  |

### 梯度变量（Map Token）

> 继承所有 SeedToken 的属性

Map Token 是基于 Seed 派生的梯度变量。定制 Map Token 推荐通过 theme.algorithm 来实现。

### 别名变量（Alias Token）

> 继承所有 MapToken 的属性

Alias Token 是 Map Token 的语意化别名，用于组件内使用。

### 模式算法（algorithm)

用于将 Seed Token 展开为 Map Token。提供了 3 种算法：

- `default`: 默认算法
- `compact`: 紧凑算法
- `loose`: 宽松算法（适用于“长辈模式”等场景）
