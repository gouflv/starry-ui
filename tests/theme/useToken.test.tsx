import { DesignTokenProvider } from '@starry-ui/theme'
import { mount } from '@vue/test-utils'
import { expect, test } from 'vitest'
import TokenRender from './components/TokenRender.vue'

test('should TokenRender works', () => {
  const vm = mount(TokenRender)
  expect(vm.html()).toMatchSnapshot()
})

test('should ThemeProvider works', () => {
  const app = {
    components: { DesignTokenProvider, TokenRender },
    template: `
      <DesignTokenProvider :token="{ colorPrimary: 'red' }">
        <TokenRender />
      </DesignTokenProvider>
    `
  }
  const vm = mount(app)
  expect(vm.text().includes('colorPrimary:red')).toBe(true)
})

test('should ThemeProvider works in nesting', () => {
  const app = {
    components: { DesignTokenProvider, TokenRender },
    template: `
      <DesignTokenProvider :token="{ colorPrimary: 'red' }">
        <DesignTokenProvider>
          <TokenRender />
        </DesignTokenProvider>
      </DesignTokenProvider>
    `
  }
  const vm = mount(app)
  expect(vm.text().includes('colorPrimary:red')).toBe(false)
})
