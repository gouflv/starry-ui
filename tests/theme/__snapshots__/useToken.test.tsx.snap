// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should TokenRender works 1`] = `
"blue:#3A8BFF;purple:#722ED1;cyan:#13C2C2;green:#52C41A;magenta:#EB2F96;pink:#eb2f96;red:#F5222D;orange:#FA8C16;yellow:#FADB14;volcano:#FA541C;geekblue:#2F54EB;gold:#FAAD14;lime:#A0D911;colorPrimary:#3a8bff;colorSuccess:#52c41a;colorWarning:#faad14;colorError:#ff4d4f;colorInfo:#3a8bff;colorTextBase:;colorBgBase:;fontFamily:-apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, 'Helvetica Neue', Arial,
'Noto Sans', sans-serif, 'Apple Color Emoji', 'Segoe UI Emoji', 'Segoe UI Symbol',
'Noto Color Emoji';fontSize:14;lineWidth:1;lineType:solid;borderRadius:6;sizeUnit:4;sizeStep:4;sizePopupArrow:16;controlHeight:32;zIndexBase:0;zIndexPopupBase:1000;motionUnit:0.1;motionBase:0;blue-1:#f0f9ff;blue-2:#def0ff;blue-3:#b5dbff;blue-4:#8cc4ff;blue-5:#63a9ff;blue-6:#3a8bff;blue-7:#276bd9;blue-8:#174eb3;blue-9:#0b348c;blue-10:#072266;purple-1:#f9f0ff;purple-2:#efdbff;purple-3:#d3adf7;purple-4:#b37feb;purple-5:#9254de;purple-6:#722ed1;purple-7:#531dab;purple-8:#391085;purple-9:#22075e;purple-10:#120338;cyan-1:#e6fffb;cyan-2:#b5f5ec;cyan-3:#87e8de;cyan-4:#5cdbd3;cyan-5:#36cfc9;cyan-6:#13c2c2;cyan-7:#08979c;cyan-8:#006d75;cyan-9:#00474f;cyan-10:#002329;green-1:#f6ffed;green-2:#d9f7be;green-3:#b7eb8f;green-4:#95de64;green-5:#73d13d;green-6:#52c41a;green-7:#389e0d;green-8:#237804;green-9:#135200;green-10:#092b00;magenta-1:#fff0f6;magenta-2:#ffd6e7;magenta-3:#ffadd2;magenta-4:#ff85c0;magenta-5:#f759ab;magenta-6:#eb2f96;magenta-7:#c41d7f;magenta-8:#9e1068;magenta-9:#780650;magenta-10:#520339;pink-1:#fff0f6;pink-2:#ffd6e7;pink-3:#ffadd2;pink-4:#ff85c0;pink-5:#f759ab;pink-6:#eb2f96;pink-7:#c41d7f;pink-8:#9e1068;pink-9:#780650;pink-10:#520339;red-1:#fff1f0;red-2:#ffccc7;red-3:#ffa39e;red-4:#ff7875;red-5:#ff4d4f;red-6:#f5222d;red-7:#cf1322;red-8:#a8071a;red-9:#820014;red-10:#5c0011;orange-1:#fff7e6;orange-2:#ffe7ba;orange-3:#ffd591;orange-4:#ffc069;orange-5:#ffa940;orange-6:#fa8c16;orange-7:#d46b08;orange-8:#ad4e00;orange-9:#873800;orange-10:#612500;yellow-1:#feffe6;yellow-2:#ffffb8;yellow-3:#fffb8f;yellow-4:#fff566;yellow-5:#ffec3d;yellow-6:#fadb14;yellow-7:#d4b106;yellow-8:#ad8b00;yellow-9:#876800;yellow-10:#614700;volcano-1:#fff2e8;volcano-2:#ffd8bf;volcano-3:#ffbb96;volcano-4:#ff9c6e;volcano-5:#ff7a45;volcano-6:#fa541c;volcano-7:#d4380d;volcano-8:#ad2102;volcano-9:#871400;volcano-10:#610b00;geekblue-1:#f0f5ff;geekblue-2:#d6e4ff;geekblue-3:#adc6ff;geekblue-4:#85a5ff;geekblue-5:#597ef7;geekblue-6:#2f54eb;geekblue-7:#1d39c4;geekblue-8:#10239e;geekblue-9:#061178;geekblue-10:#030852;gold-1:#fffbe6;gold-2:#fff1b8;gold-3:#ffe58f;gold-4:#ffd666;gold-5:#ffc53d;gold-6:#faad14;gold-7:#d48806;gold-8:#ad6800;gold-9:#874d00;gold-10:#613400;lime-1:#fcffe6;lime-2:#f4ffb8;lime-3:#eaff8f;lime-4:#d3f261;lime-5:#bae637;lime-6:#a0d911;lime-7:#7cb305;lime-8:#5b8c00;lime-9:#3f6600;lime-10:#254000;colorText:rgba(0, 0, 0, 0.88);colorTextSecondary:rgba(0, 0, 0, 0.65);colorTextTertiary:rgba(0, 0, 0, 0.45);colorTextQuaternary:rgba(0, 0, 0, 0.25);colorFill:rgba(0, 0, 0, 0.15);colorFillSecondary:rgba(0, 0, 0, 0.06);colorFillTertiary:rgba(0, 0, 0, 0.04);colorFillQuaternary:rgba(0, 0, 0, 0.02);colorBgLayout:#f5f5f5;colorBgContainer:#ffffff;colorBgElevated:#ffffff;colorBgSpotlight:rgba(0, 0, 0, 0.85);colorBorder:#d9d9d9;colorBorderSecondary:#f0f0f0;colorPrimaryBg:#f0f9ff;colorPrimaryBgHover:#def0ff;colorPrimaryBorder:#b5dbff;colorPrimaryBorderHover:#8cc4ff;colorPrimaryHover:#63a9ff;colorPrimaryActive:#276bd9;colorPrimaryTextHover:#63a9ff;colorPrimaryText:#3a8bff;colorPrimaryTextActive:#276bd9;colorSuccessBg:#f6ffed;colorSuccessBgHover:#d9f7be;colorSuccessBorder:#b7eb8f;colorSuccessBorderHover:#95de64;colorSuccessHover:#95de64;colorSuccessActive:#389e0d;colorSuccessTextHover:#73d13d;colorSuccessText:#52c41a;colorSuccessTextActive:#389e0d;colorErrorBg:#fff2f0;colorErrorBgHover:#fff1f0;colorErrorBorder:#ffccc7;colorErrorBorderHover:#ffa39e;colorErrorHover:#ff7875;colorErrorActive:#d9363e;colorErrorTextHover:#ff7875;colorErrorText:#ff4d4f;colorErrorTextActive:#d9363e;colorWarningBg:#fffbe6;colorWarningBgHover:#fff1b8;colorWarningBorder:#ffe58f;colorWarningBorderHover:#ffd666;colorWarningHover:#ffd666;colorWarningActive:#d48806;colorWarningTextHover:#ffc53d;colorWarningText:#faad14;colorWarningTextActive:#d48806;colorInfoBg:#f0f9ff;colorInfoBgHover:#def0ff;colorInfoBorder:#b5dbff;colorInfoBorderHover:#8cc4ff;colorInfoHover:#8cc4ff;colorInfoActive:#276bd9;colorInfoTextHover:#63a9ff;colorInfoText:#3a8bff;colorInfoTextActive:#276bd9;colorBgMask:rgba(0, 0, 0, 0.45);colorWhite:#fff;fontSizeSM:12;fontSizeLG:16;fontSizeXL:20;fontSizeHeading1:38;fontSizeHeading2:30;fontSizeHeading3:24;fontSizeHeading4:20;fontSizeHeading5:16;lineHeight:1.5714285714285714;lineHeightLG:1.5;lineHeightSM:1.6666666666666667;lineHeightHeading1:1.2105263157894737;lineHeightHeading2:1.2666666666666666;lineHeightHeading3:1.3333333333333333;lineHeightHeading4:1.4;lineHeightHeading5:1.5;sizeXXL:48;sizeXL:32;sizeLG:24;sizeMD:20;sizeMS:16;size:16;sizeSM:12;sizeXS:8;sizeXXS:4;controlHeightSM:24;controlHeightXS:16;controlHeightLG:40;lineWidthBold:2;borderRadiusXS:2;borderRadiusSM:4;borderRadiusLG:8;borderRadiusOuter:4;colorLink:#3a8bff;colorLinkHover:#8cc4ff;colorLinkActive:#276bd9;colorFillContent:rgba(0, 0, 0, 0.06);colorFillContentHover:rgba(0, 0, 0, 0.15);colorFillAlter:rgba(0, 0, 0, 0.02);colorBgContainerDisabled:rgba(0, 0, 0, 0.04);colorBorderBg:#ffffff;colorSplit:rgba(5, 5, 5, 0.06);colorTextPlaceholder:rgba(0, 0, 0, 0.25);colorTextDisabled:rgba(0, 0, 0, 0.25);colorTextHeading:rgba(0, 0, 0, 0.88);colorTextLabel:rgba(0, 0, 0, 0.65);colorTextDescription:rgba(0, 0, 0, 0.45);colorTextLightSolid:#fff;colorHighlight:#ff4d4f;colorBgTextHover:rgba(0, 0, 0, 0.06);colorBgTextActive:rgba(0, 0, 0, 0.15);colorIcon:rgba(0, 0, 0, 0.45);colorIconHover:rgba(0, 0, 0, 0.88);colorErrorOutline:rgba(255, 38, 5, 0.06);colorWarningOutline:rgba(255, 215, 5, 0.1);fontSizeIcon:12;controlOutlineWidth:2;controlInteractiveSize:16;controlItemBgHover:rgba(0, 0, 0, 0.04);controlItemBgActive:#f0f9ff;controlItemBgActiveHover:#def0ff;controlItemBgActiveDisabled:rgba(0, 0, 0, 0.15);controlTmpOutline:rgba(0, 0, 0, 0.02);controlOutline:rgba(5, 155, 255, 0.06);fontWeightStrong:600;opacityLoading:0.65;linkDecoration:none;linkHoverDecoration:none;linkFocusDecoration:none;controlPaddingHorizontal:12;controlPaddingHorizontalSM:8;paddingXXS:4;paddingXS:8;paddingSM:12;padding:16;paddingMD:20;paddingLG:24;paddingXL:32;paddingContentHorizontalLG:24;paddingContentVerticalLG:12;paddingContentHorizontal:12;paddingContentVertical:12;paddingContentHorizontalSM:16;paddingContentVerticalSM:8;marginXXS:4;marginXS:8;marginSM:12;margin:16;marginMD:20;marginLG:24;marginXL:32;marginXXL:48;boxShadow:
0 1px 2px 0 rgba(0, 0, 0, 0.03),
0 1px 6px -1px rgba(0, 0, 0, 0.02),
0 2px 4px 0 rgba(0, 0, 0, 0.02)
;boxShadowSecondary:
0 6px 16px 0 rgba(0, 0, 0, 0.08),
0 3px 6px -4px rgba(0, 0, 0, 0.12),
0 9px 28px 8px rgba(0, 0, 0, 0.05)
;boxShadowTertiary:
0 1px 2px 0 rgba(0, 0, 0, 0.03),
0 1px 6px -1px rgba(0, 0, 0, 0.02),
0 2px 4px 0 rgba(0, 0, 0, 0.02)
;screenXS:480;screenXSMin:480;screenXSMax:575;screenSM:576;screenSMMin:576;screenSMMax:767;screenMD:768;screenMDMin:768;screenMDMax:991;screenLG:992;screenLGMin:992;screenLGMax:1199;screenXL:1200;screenXLMin:1200;screenXLMax:1599;screenXXL:1600;screenXXLMin:1600;screenXXLMax:1999;screenXXXL:2000;screenXXXLMin:2000;boxShadowPopoverArrow:3px 3px 7px rgba(0, 0, 0, 0.1);boxShadowCard:
0 1px 2px -2px rgba(0, 0, 0, 0.16),
0 3px 6px 0 rgba(0, 0, 0, 0.12),
0 5px 12px 4px rgba(0, 0, 0, 0.09)
;boxShadowDrawerRight:
-6px 0 16px 0 rgba(0, 0, 0, 0.08),
-3px 0 6px -4px rgba(0, 0, 0, 0.12),
-9px 0 28px 8px rgba(0, 0, 0, 0.05)
;boxShadowDrawerLeft:
6px 0 16px 0 rgba(0, 0, 0, 0.08),
3px 0 6px -4px rgba(0, 0, 0, 0.12),
9px 0 28px 8px rgba(0, 0, 0, 0.05)
;boxShadowDrawerUp:
0 6px 16px 0 rgba(0, 0, 0, 0.08),
0 3px 6px -4px rgba(0, 0, 0, 0.12),
0 9px 28px 8px rgba(0, 0, 0, 0.05)
;boxShadowDrawerDown:
0 -6px 16px 0 rgba(0, 0, 0, 0.08),
0 -3px 6px -4px rgba(0, 0, 0, 0.12),
0 -9px 28px 8px rgba(0, 0, 0, 0.05)
;boxShadowTabsOverflowLeft:inset 10px 0 8px -8px rgba(0, 0, 0, 0.08);boxShadowTabsOverflowRight:inset -10px 0 8px -8px rgba(0, 0, 0, 0.08);boxShadowTabsOverflowTop:inset 0 10px 8px -8px rgba(0, 0, 0, 0.08);boxShadowTabsOverflowBottom:inset 0 -10px 8px -8px rgba(0, 0, 0, 0.08);"
`;
